crypto/bindings/conversion.ts

```diff
@@ -98,3 +158,120 @@ function fieldsFromRustFlat(fieldBytes: Uint8Array): MlArray<Field> {
   }
   return [0, ...fields];
 }
+
+// affine
+
+type WasmAffine = wasmNamespace.WasmGVesta | wasmNamespace.WasmGPallas;
+type MakeAffine<A extends WasmAffine> = () => A & { ptr: number };
+
+function affineFromRust(pt: WasmAffine): OrInfinity {
+  if (pt.infinity) {
+    pt.free();
+    return 0;
+  } else {
+    let x = fieldFromRust(pt.x);
+    let y = fieldFromRust(pt.y);
+    pt.free();
+    return [0, [0, x, y]];
+  }
+}
+
+function affineToRust<A extends WasmAffine>(
+  pt: OrInfinity,
+  klass: MakeAffine<A>
+) {
+  var res = klass();
+  if (pt === 0) {
+    res.infinity = true;
+  } else {
+    res.x = fieldToRust(pt[1][1]);
+    res.y = fieldToRust(pt[1][2]);
+  }
+  return res;
+}
+
+// polycomm
+
+type WasmPolyComm = wasmNamespace.WasmFpPolyComm | wasmNamespace.WasmFqPolyComm;
+type WasmPolyCommClass = wasm['WasmFpPolyComm'] | wasm['WasmFqPolyComm'];
+
+function polyCommFromRust(
+  polyComm: WasmPolyComm,
+  klass: WrapperClass<WasmAffine>,
+  shouldFree: boolean
+): PolyComm {
+  let rustShifted = polyComm.shifted;
+  let rustUnshifted = polyComm.unshifted;
+  let mlShifted: MlOption<OrInfinity> =
+    rustShifted === undefined ? 0 : [0, affineFromRust(rustShifted)];
+  let mlUnshifted = mlArrayFromRustVector(
+    rustUnshifted,
+    klass,
+    affineFromRust,
+    shouldFree
+  );
+  return [0, mlUnshifted, mlShifted];
+}
+
+function polyCommToRust(
+  [, camlUnshifted, camlShifted]: PolyComm,
+  PolyComm: WasmPolyCommClass,
+  makeAffine: MakeAffine<WasmAffine>
+): WasmPolyComm {
+  let rustShifted =
+    camlShifted === 0 ? undefined : affineToRust(camlShifted[1], makeAffine);
+  let rustUnshifted = mlArrayToRustVector(
+    camlUnshifted,
+    affineToRust,
+    makeAffine
+  );
+  return new PolyComm(rustUnshifted, rustShifted);
+}
+
+// generic rust helpers
+
+type Freeable = { free(): void };
+type WrappedPointer = Freeable & { ptr: number };
+type WrapperClass<T extends Freeable> = {
+  __wrap(i: number): T;
+};
+
+const registry = new FinalizationRegistry((ptr: WrappedPointer) => {
+  ptr.free();
+});
+
+function mlArrayFromRustVector<TRust extends Freeable, TMl>(
+  rustVector: Uint32Array,
+  klass: WrapperClass<TRust>,
+  convert: (c: TRust) => TMl,
+  shouldFree: boolean
+): MlArray<TMl> {
+  var n = rustVector.length;
+  var array: TMl[] = new Array(n);
+  for (let i = 0; i < n; i++) {
+    var rustValue = klass.__wrap(rustVector[i]);
+    array[i] = convert(rustValue);
+    if (shouldFree) rustValue.free();
+  }
+  return [0, ...array];
+}
+
+// TODO get rid of excessive indirection here
+
+function mlArrayToRustVector<TRust extends WrappedPointer, TMl>(
+  [, ...array]: MlArray<TMl>,
```

Comment:
Do you know how it is translated at runtime? The first element is dropped? What is pushed on the stack when calling this function? `ptr` or `ptr + single_element_size`?

Reply:
you call the function with the full array. this statement here happens inside the function. it destructures the full array to drop the first element. so we create a copy which has 1 element less.

again, I want to avoid the error-prone inline index-shifting

Reply:
Interesting that it is a two steps. Not very good in terms of runtime cost to destructure then. I didn't know.
IMHO, we should avoid it as often as we can to optimise the code. We could have sometimes large arrays and copies are not as free as we can think. And it we can avoid it... why not :).

Reply:
I opened https://github.com/o1-labs/o1js-bindings/issues/155.